src/java.base/share/classes/jdk/internal/math/ToDecimal.java

```diff
@@ -0,0 +1,184 @@
+            str[index + 1] = (byte) c2;
+        } else {
+            JLA.putCharUTF16(str, index    , c1);
+            JLA.putCharUTF16(str, index + 1, c2);
+        }
+    }
+
+    final void putDigit(byte[] str, int index, int d) {
+        putChar(str, index, (byte) ('0' + d));
+    }
+
+    final void put8Digits(byte[] str, int index, int m) {
+        /*
+         * Left-to-right digits extraction:
+         * algorithm 1 in [3], with b = 10, k = 8, n = 28.
+         */
+        if (coder == LATIN1) {
+            put8DigitsLatin1(str, index, m);
+        } else {
+            put8DigitsUTF16 (str, index, m);
+        }
+    }
+
+    private static void put8DigitsLatin1(byte[] str, int index, int m) {
+        int y = y(m);
+        for (int i = 0; i < 8; ++i) {
+            int t = 10 * y;
+            str[index + i] = (byte) ('0' + (t >>> 28));
+            y = t & MASK_28;
+        }
+    }
+
+    private static void put8DigitsUTF16(byte[] str, int index, int m) {
+        int y = y(m);
+        for (int i = 0; i < 8; ++i) {
+            int t = 10 * y;
+            JLA.putCharUTF16(str, index + i, '0' + (t >>> 28));
+            y = t & MASK_28;
+        }
+    }
+
+    static int y(int a) {
+        /*
+         * Algorithm 1 in [3] needs computation of
+         *     floor((a + 1) 2^n / b^k) - 1
+         * with a < 10^8, b = 10, k = 8, n = 28.
+         * Noting that
+         *     (a + 1) 2^n <= 10^8 2^28 < 10^17
+         * For n = 17, m = 8 the table in section 10 of [1] leads to:
+         */
+        return (int) (multiplyHigh(
+                (long) (a + 1) << 28,
+                193_428_131_138_340_668L) >>> 20) - 1;
+    }
+
+    final int removeTrailingZeroes(byte[] str, int index) {
+        if (coder == LATIN1) {
+            while (str[index - 1] == '0') {
+                --index;
+            }
+            /* ... but do not remove the one directly to the right of '.' */
+            if (str[index - 1] == '.') {
+                ++index;
+            }
+        } else {
+            while (JLA.getUTF16Char(str, index - 1) == '0') {
+                --index;
+            }
+            /* ... but do not remove the one directly to the right of '.' */
+            if (JLA.getUTF16Char(str, index - 1) == '.') {
+                ++index;
+            }
+        }
+        return index;
+    }
+
+    final int putSpecial(byte[] str, int index, int type) {
+        String s = special(type);
+        int length = s.length();
+        if (coder == LATIN1) {
+            for (int i = 0; i < length; ++i) {
+                str[index + i] = (byte) s.charAt(i);
+            }
+        } else {
+            for (int i = 0; i < length; ++i) {
+                putChar(str, index + i, s.charAt(i));
+            }
+        }
+        return index + length;
+    }
+
+    int length(byte[] str) {
+        return str.length >> coder;
+    }
+
+    /* Using the deprecated constructor enhances performance */
```

Comment:
Enhances performance over what, `new String(str, 0, index, StandardCharsets.US_ASCII)`? Please specify.

Reply:
The previous version of the code is like this, which will affect Float.toString and Double.toString. If changed, it is necessary to add the Benchmark of Float.toString and Double.toString, so I don’t want to change this in this PR.

Reply:
I followed your suggestion and changed it, but Coding uses ISO_8859_1. The performance of ISO_8859_1 is faster than that of the previously abandoned method.